src/platform/backends/lxd/lxd_vm_image_vault.cpp

```diff
@@ -207,18 +207,46 @@ mp::VMImage mp::LXDVMImageVault::fetch_image(const FetchType& fetch_type, const
         // Image doesn't exist, so move on
     }
 
-    // TODO: Remove once we do support http & file based images
-    if (query.query_type != Query::Type::Alias)
-        throw std::runtime_error("http and file based images are not supported");
-
-    const auto info = info_for(query);
-    const auto id = info.id;
     VMImage source_image;
+    VMImageInfo info;
+    QString id;
+
+    if (query.query_type == Query::Type::Alias)
+    {
+        info = info_for(query);
+        id = info.id;
+
+        source_image.id = id.toStdString();
+        source_image.original_release = info.release_title.toStdString();
+        source_image.release_date = info.version.toStdString();
+        source_image.aliases = copy_aliases(info.aliases);
+    }
+    else
+    {
+        QUrl image_url(QString::fromStdString(query.release));
+        QDateTime last_modified;
+
+        if (query.query_type == Query::Type::HttpDownload)
+        {
+            // Generate a sha256 hash based on the URL and use that for the id
+            id = QString(QCryptographicHash::hash(query.release.c_str(), QCryptographicHash::Sha256).toHex());
+            last_modified = url_downloader->last_modified(image_url);
+        }
+        else
+        {
+            if (!QFile::exists(image_url.path()))
+                throw std::runtime_error(fmt::format("Custom image `{}` does not exist.", image_url.path()));
 
-    source_image.id = id.toStdString();
-    source_image.original_release = info.release_title.toStdString();
-    source_image.release_date = info.version.toStdString();
-    source_image.aliases = copy_aliases(info.aliases);
+            source_image.image_path = image_url.path();
+            id = mp::vault::compute_image_hash(source_image.image_path);
+            last_modified = QDateTime::currentDateTime();
```

Comment:
So we're hashing the whole file? Is there value in that, vs doing what you do for http above?

Shouldn't this be file modification time? Or would it prevent us from re-fetching the image?

Reply:
What I'm trying to accomplish here (and [here](https://github.com/canonical/multipass/blob/2fc3442ad1f4d622040cde34733969006b02cc26/src/daemon/default_vm_image_vault.cpp#L309)) is that we know the hash of the image itself so in the case that someone moved the image file on their disk, we would still know it's the same image and thus not do another import in the image vault.

As far as `last_modified`, this is really only used for the `release_date` field in `VMImageInfo` and is unused at this time.